src/main/java/org/candlepin/resource/ConsumerResource.java

```diff
@@ -1608,43 +1614,48 @@ else if (toUpdate.getContentAccessMode() != null) {
      * @return boolean status whether any updates are made or not.
      */
     private boolean updateEnvironment(ConsumerDTO updated, Consumer toUpdate) {
-        if (updated.getEnvironment() == null) {
+        List<EnvironmentDTO> environments = updated.getEnvironments();
+
+        if (environments == null) {
             return false;
         }
-
         boolean changesMade = false;
-        String environmentId = updated.getEnvironment() == null ? null : updated.getEnvironment().getId();
-        String environmentName = updated.getEnvironment() == null ? null : updated.getEnvironment().getName();
-        String validatedEnvId = null;
+        Map<String, String> validatedEnvIds = new HashMap<>();
 
-        if (environmentId != null) {
-            if (toUpdate.getEnvironmentId() == null || !toUpdate.getEnvironmentId().equals(environmentId)) {
+        if (environments != null) {
+            for (EnvironmentDTO environment : environments) {
+                String environmentId = environment == null ? null : environment.getId();
+                String environmentName = environment == null ? null : environment.getName();
 
-                if (!environmentCurator.exists(environmentId)) {
-                    throw new NotFoundException(i18n.tr(
-                        "Environment with ID \"{0}\" could not be found.", environmentId));
+                if (environmentId != null) {
+                    if (!environmentCurator.exists(environmentId)) {
```

Comment:
We may want to consider bulking these calls up into a single query. Do we have an upper limit on the number of environments a consumer is allowed to join? Do we have an expectation of the average number of environments per consumer?

These may be @jturel questions, but should be considered when doing one-off queries in loops.

Reply:
My assumption would be the average would be in the single digits, but I shouldn't be assuming; we should make sure to find out about these first :+1: